src/main/java/edu/kit/compiler/register_allocation/LinearScan.java

```diff
@@ -0,0 +1,427 @@
+                        }
+                    }
+                    case DIV, MOD -> {
+                        int dividend = instr.inputRegister(0);
+                        int divisor = instr.inputRegister(1);
+                        boolean freeRAX = !analysis.getLifetime(dividend).isLastInstructionAndInput(i) ||
+                                assignment[dividend].isSpilled() || assignment[dividend].getRegister().get() != Register.RAX;
+                        if (freeRAX) {
+                            state.assertFree(Register.RAX);
+                        }
+                        state.assertFree(Register.RDX);
+                        if (assignment[divisor].isSpilled()) {
+                            state.assertCapacity(freeRAX ? 3 : 2, Optional.of(Register.RAX));
+                        }
+                        state.leaveInstruction(i);
+                        allocateTargetRegister(state, i, instr);
+                    }
+                    case MOV_S, MOV_U -> {
+                        int source = instr.inputRegister(0);
+                        int target = instr.getTargetRegister().get();
+                        boolean isSignedUpcast = sizes[target].getBytes() > sizes[source].getBytes()
+                                && instr.getType() == InstructionType.MOV_S;
+                        if ((!assignment[source].isInRegister() || isSignedUpcast) &&
+                                !assignment[target].isInRegister()) {
+                            state.assertCapacity(1, Optional.empty());
+                        }
+                        state.leaveInstruction(i);
+                        allocateTargetRegister(state, i, instr);
+                    }
+                    case CALL, RET -> {
+                        state.leaveInstruction(i);
+                        allocateTargetRegister(state, i, instr);
+                    }
+                }
+
+                i++;
+            }
+        }
+        state.assertFinallyEmpty(i);
+        state.calculateStackSlots();
+
+        return ApplyAssignment.createFunctionBody(
+                assignment, sizes, analysis.getLifetimes(), input, analysis.getNumInstructions(), nArgs);
+    }
+
+    private void allocateTargetRegister(ScanState state, int index, Instruction instr) {
+        if (instr.getTargetRegister().isPresent()) {
+            LifetimeAnalysis analysis = state.getAnalysis();
+            int target = instr.getTargetRegister().get();
+            if (analysis.getLifetime(target).getBegin() == index) {
+                // allocate the new register
+                RegisterPreference preference = calculatePreference(target, state.getAssignment(), analysis);
+                state.allocateRegister(target, preference);
+            }
+        }
+    }
+
+    private static RegisterPreference calculatePreference(int vRegister, RegisterAssignment[] assignment,
+                                                          LifetimeAnalysis analysis) {
```

Comment:
More of a general point, there are a excessively long methods in this class, as well as in `ApplyAssignment`. Maybe some of these methods could be broken into multiple logic units, to improve readability and possibly find some opportunities for code-reuse.

Reply:
I have looked into it and found a few minor things and one bigger block that could be refactored as logically separate units.

However, all methods are/were below or around 100 lines, so I wouldn't call them particularly long. I don't think introducing new methods only for the sake of reducing line counts is a good idea here.

Reply:
The changes you made look good. If there is no real opportunity for further code reuse, you're probably right and we can leave it as it is.